### PR TITLE
feat: Improved compliance of Encoding APIs

### PR DESCRIPTION
**Pull request by nabetti1720**
### Description of changes

Add the following methods/properties to improve the compliance of Encoding API.

- TextEncoder.encoding
- TextDecoder.fatal
- TextDecoder.ignoreBOM
- TextDecoder.encoding
- TextDecoder.decode()

And the `encoding_rs` crate was adopted for simplicity in dealing with various labels of `TextDecoder()`.
 
### Checklist

- [x] Created unit tests in `tests/unit` and/or in Rust for my feature if needed
- [x] Ran `make fix` to format JS and apply Clippy auto fixes
- [x] Made sure my code didn't add any additional warnings: `make check`
- [x] Updated documentation if needed ([API.md](API.md)/[README.md](README.md)/Other)

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: llrt_core/src/modules/encoding/string_builder.rs
Comment (richarddavison) [inline llrt_core/src/modules/encoding/string_builder.rs:1]: I think we can remove this file, it's not being used anywhere

File: llrt_core/src/modules/encoding/text_decoder.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    encoding: String,
+    fatal: bool,
+    ignore_bom: bool,
+}
+
+#[rquickjs::methods]
+impl<'js> TextDecoder {
+    #[qjs(constructor)]
+    pub fn new(ctx: Ctx<'js>, label: Opt<String>, options: Opt<Object<'js>>) -> Result<Self> {
+        let label = match label.0 {
```
Comment (richarddavison) [inline llrt_core/src/modules/encoding/text_decoder.rs:23]:
```rust
let label = label.0
    .filter(|lbl| !lbl.is_empty())
    .unwrap_or_else(|| String::from("utf-8"));
```

File: llrt_core/src/modules/encoding/text_decoder.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            Some(lbl) if !lbl.is_empty() => lbl,
+            Some(_) | None => String::from("utf-8"),
+        };
+        let mut fatal = false;
+        let mut ignore_bom = false;
+
+        let encoding = match Encoding::for_label(label.as_bytes()) {
```
Comment (richarddavison) [inline llrt_core/src/modules/encoding/text_decoder.rs:30]: Simplify:
```rust
let encoding = Encoding::for_label(label.as_bytes())
    .map(|enc| enc.name().to_string()).or_throw_message(&ctx,"Unsupported encoding label")
```

File: llrt_core/src/modules/encoding/text_decoder.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        if self.fatal && has_error {
+            return Err(Exception::throw_message(&ctx, "Fatal error"));
+        }
+
+        Ok(str.into_owned())
+    }
+}
+
+fn get_option<'js, V: FromJs<'js> + Sized>(
```
Comment (richarddavison) [inline llrt_core/src/modules/encoding/text_decoder.rs:90]: Already in utils as  `get_optional`

**Review by richarddavison — CHANGES_REQUESTED** [review]
Looks very good! Just some minor things :)

**Comment by nabetti1720** [comment]
Hi @richarddavison , Thank you for your always accurate reviews.
All have been corrected, please confirm. :)

**Pull request merged**